tool/synthetic-data-sampler.ts

```diff
@@ -0,0 +1,530 @@
+    let headings : string = '';
+    if (useHeading)
+        headings = ["id", "utterance", "thingtalk", "query", "queryCanonical", "argument", "value"].join('\t') + '\n';
+    const rows = data.reduce((acc : string[], colValue, idx) => {
+        const id = `${device}-${idx.toLocaleString('en-US', {minimumIntegerDigits: 3, useGrouping:false})}`;
+        const tmp = [
+            id,
+            colValue["utterance"],
+            colValue["thingtalk"]
+        ]
+        return acc.concat([tmp.join('\t')]);
+    }, []).join('\n');
+    return `${headings}${rows}`;
+}
+
+function toThingtalkValue(value : string, type : Type) : { value : Ast.Value, op : string } {
+    value = value.toLowerCase();
+    if (type instanceof Type.Entity) {
+        value = `"${value}"`;
+        return {value: new Ast.Value.Entity(value, 'tt:device', value), op: "=~" };
+    }
+    if (type instanceof Type.Enum) {
+        return { value: new Ast.Value.Enum(value), op: "==" };
+    }
+    if (type === Type.String) {
+        // value = `"${value}"`;
+        // return {value: new Ast.Value.Entity(value, 'tt:device', value), op: "=~" };
+        return { value: new Ast.Value.String(value), op: "=~" };
+    }
+    if (type === Type.Number) {
+        return { value: new Ast.Value.Number(parseFloat(value)), op: "==" };
+    }
+    if (type instanceof Type.Array) {
+        type = type.elem as Type;
+        if (type instanceof Type.Entity) {
+            value = `"${value}"^^${type.type}("${value}")`;
+            return { value: new Ast.Value.Entity(value, 'tt:device', value), op: "contains" };
+        } else {
+            throw new Error(`Unsupported value type: ${type}`);
+        }
+    }
+    throw new Error(`Unsupported value type: ${type}`);
+}
+
+function checkOrCreateOutputPath(args : any) {
+    if (!args.output) {
+        const outputDir = path.join(path.dirname(args.thingpedia), "test");
+        if (!fs.existsSync(outputDir)) {
+            fs.mkdirSync(outputDir);
+        }
+        args.output = fs.createWriteStream(path=path.join(outputDir, "results.tsv"));
+    }
+}
+
+export async function execute(args : any) {
+    process.stdout.write("Generating samples... ");
+    checkOrCreateOutputPath(args);
+    const tpClient = new Tp.FileClient(args);
+    const schemaRetriever = new SchemaRetriever(tpClient, null, !args.debug);
+    const device = args.devices.split(',')[0];
+    const locale = args.locale;
+    const sampleSize = args.sample_size;
+    const deviceClass = await schemaRetriever.getFullSchema(device);
+    const baseTokenizer : I18n.BaseTokenizer = I18n.get(args.locale).getTokenizer(); 
+    // const functions = Object.assign({}, deviceClass.queries, deviceClass.actions);
+    const functions = Object.assign({}, deviceClass.queries);
+    let sampledConstants : Record<string, Constant[]> = {}; 
+    if (args.constants) {
+        sampledConstants = await parseConstantFile(args.locale, args.constants);
+    } else {
```

Comment:
why do you want to replicate the `sample-constants` command here? If there is no particular reason, we should defintiely just run `sample-constants` first and assume constants file is available.